### ISSUE TITLE
How to type props of a route with Typescript?

### ISSUE THREAD

**Issue by eivindml** [expo/router#103]
Is there an exported type we can use to type the props? 

In Next.js this is type like:

```ts
context.params?.id // undefined | string | string[]
```

```app/blog/[id].js```
```ts
export default function BlogPost({ route }) {
  return (
    <Text>
      {route.params.id}
    </Text>
  );
}
```

**Comment by koka0012** [comment]
Expo Router is based in react-navigation, which that means you can use react-navigation props to type your rote params.
Here's an example:

https://reactnavigation.org/docs/typescript/

**Comment by karlhorky** [comment]
Would be good to get an example of usage of `NativeStackScreenProps` in a working example app eg. in [`apps/demo`](https://github.com/expo/router/tree/main/apps/demo) to see if there are any differences with the way that `expo-router` handles this vs vanilla `react-navigation`.

Eg. does this example below from the the link from @koka0012 translate 1-to-1 to usage in `expo-router`?

```tsx
import type { NativeStackScreenProps } from '@react-navigation/native-stack';

type RootStackParamList = {
  Home: undefined;
  Profile: { userId: string };
  Feed: { sort: 'latest' | 'top' } | undefined;
};

type Props = NativeStackScreenProps<RootStackParamList, 'Profile'>;
```

cc @EvanBacon

**Comment by JClackett** [comment]
Yes please to exported type for route's!

**Comment by danielkv** [comment]
Having that, directly from `expo-router` would be awesome 😎

**Comment by EvanBacon** [comment]
My focus is primarily on getting things working consistently without errors, but I would love to have some fully automated type system in the future.

Off the top, we probably need the following:

- `<Link />` typed correctly.
- Some guide on manually typing the routes for `useHref<...>()`
- Improved types for `useLink()`

Further, it would be nice to have some automatic typing (far in the future as this is fragile and strenuous to maintain):

- `useLink.replace('...')` should autocomplete the available routes.
- `<Link href />` should support autocomplete.
- Perhaps dynamic routes could be automatic.
- Could make a TypeScript service for vs code which adds autocomplete for ErrorBoundary and other exports.

@marklawlor probably has some good ideas for this.

**Comment by karlhorky** [comment]
Nice, thanks for the answer!

Do your lists above also include the route parameters, which were the things that [the original issue description](https://github.com/expo/router/issues/103#issue-1417901440) and [my comment above](https://github.com/expo/router/issues/103#issuecomment-1307566867) were referring to?

**Comment by JClackett** [comment]
Yeah I think one of the main things I needed was just a generic that would give me the props that each route has, I dont (right now) need type safe route params etc

e.g.

```ts
import { RouteProps } from "expo-router"

interface MyParams {
 id: string
}
export default function MyRoute(props: RouteProps<MyParams>) {
...
```
